include/oneapi/dpl/pstl/hetero/dpcpp/parallel_backend_sycl_utils.h

```diff
@@ -435,71 +435,28 @@ struct __sycl_usm_alloc
     }
 };
 
-//-----------------------------------------------------------------------
-// type traits for objects granting access to some value objects
-//-----------------------------------------------------------------------
-
-template <typename _ContainerOrIterator>
-struct __memobj_traits
-{
-    using value_type = typename _ContainerOrIterator::value_type;
-};
-
-template <typename _T>
-struct __memobj_traits<_T*>
-{
-    using value_type = _T;
-};
-
-} // namespace __internal
-
-template <typename _ExecutionPolicy, typename _T>
-using __buffer = __internal::__buffer_impl<::std::decay_t<_ExecutionPolicy>, _T>;
-
-template <typename T>
-struct __repacked_tuple
-{
-    using type = T;
-};
-
-template <typename... Args>
-struct __repacked_tuple<::std::tuple<Args...>>
-{
-    using type = oneapi::dpl::__internal::tuple<Args...>;
-};
-
-template <typename T>
-using __repacked_tuple_t = typename __repacked_tuple<T>::type;
-
-template <typename _ContainerOrIterable>
-using __value_t = typename __internal::__memobj_traits<_ContainerOrIterable>::value_type;
-
 template <typename _T>
 struct __accessor
 {
   private:
     using __accessor_t = sycl::accessor<_T, 1, sycl::access::mode::read_write, __dpl_sycl::__target_device,
                                         sycl::access::placeholder::false_t>;
-    __accessor_t __acc;
-    _T* __ptr = nullptr;
+
     bool __usm = false;
+    _T* __ptr = nullptr;
+    __accessor_t __acc;
 
   public:
 // A buffer is used by default. Supporting compilers use the unified future on top of USM host memory or a buffer.
 #if _ONEDPL_SYCL_USM_HOST_PRESENT
-    __accessor(sycl::handler& __cgh, bool __u, ::std::shared_ptr<sycl::buffer<_T, 1>> __sycl_buf,
-               ::std::shared_ptr<_T> __usm_buf)
-        : __usm(__u)
+    __accessor(sycl::handler& __cgh, bool __u, sycl::buffer<_T, 1>* __sycl_buf, _T* __usm_buf)
+        : __usm(__u), __ptr(__u ? __usm_buf : nullptr),
+          __acc(__u ? decltype(__acc){} : sycl::accessor(*__sycl_buf, __cgh, sycl::read_write, __dpl_sycl::__no_init{}))
     {
-        if (__usm)
-            __ptr = __usm_buf.get();
-        else
-            __acc = sycl::accessor(*__sycl_buf, __cgh, sycl::read_write, __dpl_sycl::__no_init{});
     }
```

Comment:
Same question: is it worth to change the code like this for sake of a default ctor and assignment of sycl::accessor? The logic of the code is much more obfuscated this way.

Reply:
The goal: better to initialize some smart pointers through constructor - I believe it's should work faster then assignment.